From: Bernhard Kaindl <bk@suse.de>
Subject: Allow physical DMA using OHCI-1394 early for debugging over FireWire
Patch-mainline: hopefully soon

Signed-Off-By: Bernhard Kaindl <bk@suse.de>

Automatically created from "patches.drivers/early-firewire.diff" by xen-port-patches.py

Index: head-2008-01-28/arch/x86/kernel/setup_32-xen.c
===================================================================
--- head-2008-01-28.orig/arch/x86/kernel/setup_32-xen.c	2008-01-28 13:46:37.000000000 +0100
+++ head-2008-01-28/arch/x86/kernel/setup_32-xen.c	2008-01-28 13:48:11.000000000 +0100
@@ -47,6 +47,7 @@
 #include <linux/crash_dump.h>
 #include <linux/dmi.h>
 #include <linux/pfn.h>
+#include <linux/init_ohci1394_dma.h>
 
 #include <video/edid.h>
 
@@ -719,6 +720,16 @@ void __init setup_arch(char **cmdline_p)
 	smp_alloc_memory(); /* AP processor realmode stacks in low memory*/
 #endif
 	paging_init();
+
+	/*
+	 * NOTE: On x86-32, only from this point on, fixmaps are ready for use.
+	 */
+
+#ifdef CONFIG_PROVIDE_OHCI1394_DMA_INIT
+	if (init_ohci1394_dma_early)
+		init_ohci1394_dma_on_all_controllers();
+#endif
+
 	remapped_pgdat_init();
 	sparse_init();
 	zone_sizes_init();
Index: head-2008-01-28/arch/x86/kernel/setup_64-xen.c
===================================================================
--- head-2008-01-28.orig/arch/x86/kernel/setup_64-xen.c	2008-01-28 13:36:36.000000000 +0100
+++ head-2008-01-28/arch/x86/kernel/setup_64-xen.c	2008-01-28 13:48:11.000000000 +0100
@@ -39,6 +39,7 @@
 #include <linux/dmi.h>
 #include <linux/dma-mapping.h>
 #include <linux/ctype.h>
+#include <linux/init_ohci1394_dma.h>
 
 #include <asm/mtrr.h>
 #include <asm/uaccess.h>
@@ -312,6 +313,11 @@ static void discover_ebda(void)
 #define discover_ebda() ((void)0)
 #endif
 
+/*
+ * setup_arch - architecture-specific boot-time initializations
+ *
+ * Note: On x86_64, fixmaps are ready for use even before this is called.
+ */
 void __init setup_arch(char **cmdline_p)
 {
 #ifdef CONFIG_XEN
@@ -382,6 +388,11 @@ void __init setup_arch(char **cmdline_p)
 
 	parse_early_param();
 
+#ifdef CONFIG_PROVIDE_OHCI1394_DMA_INIT
+	if (init_ohci1394_dma_early)
+		init_ohci1394_dma_on_all_controllers();
+#endif
+
 	finish_e820_parsing();
 
 	e820_register_active_regions(0, 0, -1UL);
Index: head-2008-01-28/include/asm-x86/mach-xen/asm/fixmap_32.h
===================================================================
--- head-2008-01-28.orig/include/asm-x86/mach-xen/asm/fixmap_32.h	2008-01-28 13:34:55.000000000 +0100
+++ head-2008-01-28/include/asm-x86/mach-xen/asm/fixmap_32.h	2008-01-28 13:48:11.000000000 +0100
@@ -98,6 +98,9 @@ enum fixed_addresses {
 	FIX_BTMAP_END = __end_of_permanent_fixed_addresses,
 	FIX_BTMAP_BEGIN = FIX_BTMAP_END + NR_FIX_BTMAPS - 1,
 	FIX_WP_TEST,
+#ifdef CONFIG_PROVIDE_OHCI1394_DMA_INIT
+	FIX_OHCI1394_BASE,
+#endif
 	__end_of_fixed_addresses
 };
 
Index: head-2008-01-28/include/asm-x86/mach-xen/asm/fixmap_64.h
===================================================================
--- head-2008-01-28.orig/include/asm-x86/mach-xen/asm/fixmap_64.h	2008-01-28 13:34:55.000000000 +0100
+++ head-2008-01-28/include/asm-x86/mach-xen/asm/fixmap_64.h	2008-01-28 13:48:11.000000000 +0100
@@ -46,6 +46,9 @@ enum fixed_addresses {
 	FIX_IO_APIC_BASE_0,
 	FIX_IO_APIC_BASE_END = FIX_IO_APIC_BASE_0 + MAX_IO_APICS-1,
 #endif
+#ifdef CONFIG_PROVIDE_OHCI1394_DMA_INIT
+	FIX_OHCI1394_BASE,
+#endif
 #ifdef CONFIG_ACPI
 	FIX_ACPI_BEGIN,
 	FIX_ACPI_END = FIX_ACPI_BEGIN + FIX_ACPI_PAGES - 1,
